File: 프로그래머스/level2/수식 최대화.js
let exp = (a, b, ex) => {
    if(ex==='*')        return a*b;
    else if(ex==='+')   return a+b;
    else                return a-b;
}

function solution(expression) {
    var answer = 0;
    let orders = [['+','-','*'], ['+','*','-'], ['-','+','*'], ['-','*','+'], ['*','-','+'], ['*','+','-']];
    orders.forEach(order=>{
        let numbers = expression.split(/([+\-*/])/);
        for(let i=0; i<order.length; i++){
            let ex = order[i];
            let tmp = [];
            for(let n=0; n<numbers.length; n++){
                if(numbers[n]===ex){
                    let a = Number(tmp[tmp.length-1]);
                    let b = Number(numbers[n+1]);
                    tmp.pop();
                    tmp.push(exp(a,b,numbers[n]));
                    n++;
                }
                else{
                    tmp.push(numbers[n]);
                }
            }
            numbers = tmp.slice(0);
        }
        answer = Math.max(answer, Math.abs(Number(numbers[0])));
    })
    return answer;
}

File: 프로그래머스/level2/수식최대화.cpp
//구현
#include <string>
#include <vector>
#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;

char order[][3] = {
    {'+', '-', '*'},
    {'+', '*', '-'},
    {'-', '+', '*'},
    {'-', '*', '+'},
    {'*', '+', '-'},
    {'*', '-', '+'}
};

long long exp(long long a, long long b, string ex){
    if(ex == "+")       return a+b;
    else if(ex == "-")  return a-b;
    else if(ex == "*")  return a*b;
}

long long solution(string expression) {
    long long answer = 0;
    string word = "";
    vector<string> copy;
    for(int i=0; i<expression.length(); i++){
        if(expression[i]>='0'&&expression[i]<='9'){
            word += expression[i];
        }
        else{
            copy.push_back(word);
            string f = "";
            f += expression[i];
            copy.push_back(f);
            word = "";
        }
    }
    copy.push_back(word);
    for(int i=0; i<6; i++){
        vector<string> list = copy;
        for(int j=0; j<3; j++){
            vector<string> tmp;
            string target = "";
            target += order[i][j];
            for(int k=0; k<list.size(); k++){
                if(list[k] == target){
                    long long a = stoll(tmp[tmp.size()-1]);
                    long long b = stoll(list[k+1]);
                    long long res = exp(a, b, list[k]);
                    tmp.pop_back();
                    tmp.push_back(to_string(res));
                    k++;
                }
                else{
                    tmp.push_back(list[k]);
                }
            }
            list = tmp;
        }
        answer = max(answer, abs(stoll(list[0])));
    }
    return answer;
}